authz/src/main/java/io/grpc/authz/AuthorizationPolicyTranslator.java

```diff
@@ -0,0 +1,197 @@
+ */
+public class AuthorizationPolicyTranslator {
+  private static final ImmutableList<String> UNSUPPORTED_HEADERS = ImmutableList.of(
+      "host", "connection", "keep-alive", "proxy-authenticate", "proxy-authorization",
+      "te", "trailer", "transfer-encoding", "upgrade");
+
+  private static StringMatcher getStringMatcher(String value) {
+    if (value.equals("*")) {
+      return StringMatcher.newBuilder().setPrefix("").build();
+    } else if (value.startsWith("*")) {
+      return StringMatcher.newBuilder().setSuffix(value.substring(1)).build();
+    } else if (value.endsWith("*")) {
+      return StringMatcher.newBuilder().setPrefix(value.substring(0, value.length() - 1)).build();
+    }
+    return StringMatcher.newBuilder().setExact(value).build();
+  }
+
+  private static Principal parseSource(Map<String, ?> source) {
+    List<String> principalsList = JsonUtil.getListOfStrings(source, "principals");
+    if (principalsList == null) {
+      return Principal.newBuilder().setAny(true).build();
+    }
+    if (principalsList.isEmpty()) {
+      return Principal.newBuilder().setAny(true).build();
+    }
+    Principal.Set.Builder principalsSet = Principal.Set.newBuilder();
+    for (String principal: principalsList) {           
+      principalsSet.addIds(
+          Principal.newBuilder().setAuthenticated(
+            Authenticated.newBuilder().setPrincipalName(
+              getStringMatcher(principal)).build()).build());
+    }
+    return Principal.newBuilder().setOrIds(principalsSet.build()).build();
+  }
+
+  private static Permission parseHeader(Map<String, ?> header) throws IllegalArgumentException {
+    String key = JsonUtil.getString(header, "key");
+    if (key == null) {
+      throw new IllegalArgumentException("\"key\" is absent.");
+    }
+    if (key.charAt(0) == ':'
+        || key.startsWith("grpc-")
+        || UNSUPPORTED_HEADERS.contains(key.toLowerCase())) {
+      throw new IllegalArgumentException(String.format("Unsupported \"key\" %s.", key));
+    }
+    List<String> valuesList = JsonUtil.getListOfStrings(header, "values");
+    if (valuesList == null || (valuesList != null && valuesList.isEmpty())) {
+      throw new IllegalArgumentException("\"values\" is absent or empty.");
+    }
+    Permission.Set.Builder orSet = Permission.Set.newBuilder();
+    for (String value: valuesList) {
+      orSet.addRules(
+          Permission.newBuilder().setHeader(
+            HeaderMatcher.newBuilder()
+            .setName(key)
+            .setStringMatch(getStringMatcher(value)).build()).build());     
+    }
+    return Permission.newBuilder().setOrRules(orSet.build()).build();
+  }
+
+  private static Permission parseRequest(Map<String, ?> request) throws IllegalArgumentException {
+    Permission.Set.Builder andSet = Permission.Set.newBuilder();
+    List<String> pathsList = JsonUtil.getListOfStrings(request, "paths"); 
+    if (pathsList != null) {
```

Comment:
`&& !pathList.isEmpty()` so that you don't need the `pathsSet.getRulesCount() > 0` check? It seems that'd be more clear. Ditto elsewhere.

Reply:
Done